### PR TITLE
Report project type and target framework

### PR DESCRIPTION
**Pull request by akshita31**
Created an IMSBuildSink which will be called everytime a project load completes and using this we can obtain the information regarding the project like target framework and project file path and then push the corresponding event to the event sink which will then be responded to by the C# extension

omnisharp-vscode side : https://github.com/OmniSharp/omnisharp-vscode/pull/2890

### CODE REVIEW

File: src/OmniSharp.Host/CommandLineApplication.cs [hunk 1]
```diff
@@ -80,7 +80,7 @@ private void DebugAttach()
         {
             if (Debug)
             {
-                //Console.WriteLine($"Attach debugger to process {Process.GetCurrentProcess().Id} to continue...");
+                Console.WriteLine($"Attach debugger to process {Process.GetCurrentProcess().Id} to continue...");
```
Comment (jonsequitur) [inline src/OmniSharp.Host/CommandLineApplication.cs]: Did you mean to leave this uncommented?

File: tests/OmniSharp.MSBuild.Tests/ProjectLoadListenerTests.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+            // Assert
+            Assert.Empty(tfm);
+        }
+
+        [Fact]
+        public async Task The_target_framework_is_emitted()
+        {
+            // Arrange
+            var expectedTFM = "netcoreapp1.0";
```
Comment (rchande) [inline tests/OmniSharp.MSBuild.Tests/ProjectLoadListenerTests.cs]: This should be hashed

File: src/OmniSharp.MSBuild/ProjectLoadListener.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+            catch (Exception ex)
+            {
+                _logger.LogError("Unexpected exception got thrown from project load listener: " + ex);
+            }
+        }
+
+        // Internal for testing
+        internal static string GetTargetFramework(ProjectInstance projectInstance)
+        {
+            var targetFramework = projectInstance.GetPropertyValue(TargetFrameworkPropertyName);
```
Comment (filipw) [inline src/OmniSharp.MSBuild/ProjectLoadListener.cs]: should we look at `<TargetFrameworks>` too?

**Comment by akshita31** [comment]
@rchande We need only the reference names and not the full path right ?

File: src/OmniSharp.MSBuild/HashingAlgorithm.cs [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+using System.Threading.Tasks;
+
+namespace OmniSharp.MSBuild
+{
+    public interface IHasher
+    {
+        string HashInput(string clearText);
+    }
+
+    public class VsMachashingAlgorithm : IHasher
```
Comment (rchande) [inline src/OmniSharp.MSBuild/HashingAlgorithm.cs]: `VsMacHashing`?

File: src/OmniSharp.MSBuild/ProjectLoadListener.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        internal static string GetHashedTargetFramework(ProjectInstance projectInstance)
+        {
+            var targetFramework = projectInstance.GetPropertyValue(TargetFrameworkPropertyName);
+            if (string.IsNullOrEmpty(targetFramework))
+            {
+                targetFramework = projectInstance.GetPropertyValue(TargetFrameworksPropertyName);
+            }
+            if (string.IsNullOrEmpty(targetFramework))
+            {
+                targetFramework = projectInstance.GetPropertyValue(TargetFrameworkVersionPropertyName);
```
Comment (rchande) [inline src/OmniSharp.MSBuild/ProjectLoadListener.cs]: Can this case actually happen?

File: src/OmniSharp.MSBuild/ProjectLoadListener.cs (hunk 3 above)
Comment (akshita31) [reply src/OmniSharp.MSBuild/ProjectLoadListener.cs]: Done. Thanks!

File: src/OmniSharp.MSBuild/HashingAlgorithm.cs (hunk 4 above)
Comment (akshita31) [reply src/OmniSharp.MSBuild/HashingAlgorithm.cs]: Removed

File: src/OmniSharp.MSBuild/Notification/ProjectLoadedEventArgs.cs [hunk 6]
```diff
@@ -11,17 +12,26 @@ public class ProjectLoadedEventArgs
...
+            bool isReload,
+            bool hasProjectIdGeneratedFromSolution,
+            ImmutableArray<string> sourceFiles,
+            IEnumerable<string> references = null)
         {
             Id = id;
             ProjectInstance = projectInstance;
             Diagnostics = diagnostics;
             IsReload = isReload;
+            HasProjectIdGeneratedFromSolution = hasProjectIdGeneratedFromSolution;
```
Comment (rchande) [inline src/OmniSharp.MSBuild/Notification/ProjectLoadedEventArgs.cs]: "DefinedIn" is maybe a more accurate than "GeneratedFrom"

File: src/OmniSharp.MSBuild/ProjectSystem.cs [hunk 7]
```diff
@@ -226,4 +235,17 @@ async Task<object> IProjectSystem.GetProjectModelAsync(string filePath)
...
+    public class ProjectIdInfo
+    {
+        public ProjectIdInfo(ProjectId id, bool isObtainedFromSolution)
+        {
+            Id = id ?? throw new ArgumentNullException(nameof(id));
+            IsObtainedFromSolution = isObtainedFromSolution;
+        }
+
+        public ProjectId Id { get; set; }
+        public bool IsObtainedFromSolution { get; set; }
```
Comment (rchande) [inline src/OmniSharp.MSBuild/ProjectSystem.cs]: Let's name this thing consistently

File: src/OmniSharp.MSBuild/ProjectFile/ProjectFileInfo.cs [hunk 8]
```diff
@@ -67,19 +69,21 @@ internal static ProjectFileInfo CreateEmpty(string filePath)
         {
             var id = ProjectId.CreateNewId(debugName: filePath);
 
-            return new ProjectFileInfo(id, filePath, data: null);
+            return new ProjectFileInfo(new ProjectIdInfo(id, false), filePath, data: null);
```
Comment (rchande) [inline src/OmniSharp.MSBuild/ProjectFile/ProjectFileInfo.cs]: named arguments for boolean constants

File: src/OmniSharp.MSBuild/ProjectFile/ProjectFileInfo.cs [hunk 9]
```diff
@@ -67,19 +69,21 @@ internal static ProjectFileInfo CreateEmpty(string filePath)
...
         }
 
         internal static ProjectFileInfo CreateNoBuild(string filePath, ProjectLoader loader)
         {
             var id = ProjectId.CreateNewId(debugName: filePath);
             var project = loader.EvaluateProjectFile(filePath);
             var data = ProjectData.Create(project);
+            //we are not reading the solution here 
+            var projectIdInfo = new ProjectIdInfo(id, false);
```
Comment (rchande) [inline src/OmniSharp.MSBuild/ProjectFile/ProjectFileInfo.cs]: named arg

File: src/OmniSharp.MSBuild/ProjectLoadListener.cs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+using OmniSharp.Eventing;
+using OmniSharp.Models;
+using OmniSharp.MSBuild.Notification;
+
+namespace OmniSharp.MSBuild
+{
+    [Export(typeof(IMSBuildEventSink))]
+    public class ProjectLoadListener : IMSBuildEventSink
+    {
+        internal const string TargetFrameworkPropertyName = "TargetFramework";
```
Comment (rchande) [inline src/OmniSharp.MSBuild/ProjectLoadListener.cs]: style points, you can do `internal const string TargetFramework = nameof(TargetFramework)`

File: src/OmniSharp.MSBuild/ProjectLoadListener.cs [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+            var projectFilePath = args.ProjectInstance.GetPropertyValue(MSBuildProjectFullPathPropertyName);
+            var content = File.ReadAllText(projectFilePath);
+            //create a hash from the filename and the content
+            return _referenceHashingAlgorithm.HashInput($"Filename: {Path.GetFileName(projectFilePath)}\n{content}");
+        }
+
+        private static IEnumerable<HashedString> GetHashedReferences(ProjectLoadedEventArgs args)
+        {
+            var referenceNames = args.References.Select(reference => Path.GetFileNameWithoutExtension(reference).ToLower());
```
Comment (rchande) [inline src/OmniSharp.MSBuild/ProjectLoadListener.cs]: Should we be removing the extension?

Comment (rchande) [reply src/OmniSharp.MSBuild/ProjectLoadListener.cs]: Should we be tolower?

File: src/OmniSharp.MSBuild/ProjectLoadListener.cs [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+            return referenceNames.Select(reference => _referenceHashingAlgorithm.HashInput(reference));
```
Comment (rchande) [inline src/OmniSharp.MSBuild/ProjectLoadListener.cs]: You can write `referenceNames.Select(_referenceHashingAlgorithm.HashInput)`

File: src/OmniSharp.MSBuild/ProjectSystem.cs [hunk 13]
```diff
@@ -108,50 +108,58 @@ public void Initalize(IConfiguration configuration)
...
                 _solutionFileOrRootPath = solutionFilePath;
-                return GetProjectPathsFromSolution(solutionFilePath);
+                return GetProjectPathsAndIdsFromSolution(solutionFilePath);
             }
 
             // Finally, if there isn't a single solution immediately available,
             // Just process all of the projects beneath the root path.
             _solutionFileOrRootPath = _environment.TargetDirectory;
-            return _fileSystemHelper.GetFiles("**/*.csproj");
+            var projectFilePaths = _fileSystemHelper.GetFiles("**/*.csproj");
```
Comment (rchande) [inline src/OmniSharp.MSBuild/ProjectSystem.cs]: You can use .Select on this

File: src/OmniSharp.MSBuild/ProjectLoadListener.cs (hunk 11 above)
Comment (akshita31) [reply src/OmniSharp.MSBuild/ProjectLoadListener.cs]: Yeah as per what Andrew said we need to do "system.core"

File: src/OmniSharp.MSBuild/ProjectSystem.cs (hunk 13 above)
Comment (akshita31) [reply src/OmniSharp.MSBuild/ProjectSystem.cs]: Done. Thanks!

**Pull request merged**